pkg/watermark/fetch/edge_fetcher_test.go

```diff
@@ -1342,3 +1363,400 @@ func TestFetcherWithSameOTBucket(t *testing.T) {
+			}
+		default:
+			time.Sleep(1 * time.Millisecond)
+			allProcessors = processorManager.GetAllProcessors()
+		}
+	}
+
+	// publish an idle watermark: simulate reduce
+	otValueByteOne, err = otValueToBytes(106, epoch+600, true, 0)
+	assert.NoError(t, err)
+	otValueByteTwo, err = otValueToBytes(106, epoch+600, true, 1)
+	assert.NoError(t, err)
+	otValueByteThree, err = otValueToBytes(106, epoch+600, true, 2)
+	assert.NoError(t, err)
+	err = otStore.PutKV(ctx, "p1", otValueByteOne)
+	assert.NoError(t, err)
+	err = otStore.PutKV(ctx, "p1", otValueByteTwo)
+	assert.NoError(t, err)
+	err = otStore.PutKV(ctx, "p1", otValueByteThree)
+	assert.NoError(t, err)
+
+	// p1 should get the head offset watermark from p2
+	for allProcessors["p1"].GetOffsetTimelines()[0].Dump() != "[IDLE 1651161660600:106] -> [1651161660500:103] -> [-1:-1] -> [-1:-1] -> [-1:-1] -> [-1:-1] -> [-1:-1] -> [-1:-1] -> [-1:-1] -> [-1:-1]" &&
+		allProcessors["p1"].GetOffsetTimelines()[1].Dump() != "[IDLE 1651161660600:106] -> [1651161660500:103] -> [-1:-1] -> [-1:-1] -> [-1:-1] -> [-1:-1] -> [-1:-1] -> [-1:-1] -> [-1:-1] -> [-1:-1]" &&
+		allProcessors["p1"].GetOffsetTimelines()[2].Dump() != "[IDLE 1651161660600:106] -> [1651161660500:103] -> [-1:-1] -> [-1:-1] -> [-1:-1] -> [-1:-1] -> [-1:-1] -> [-1:-1] -> [-1:-1] -> [-1:-1]" {
+
+		select {
+		case <-ctx.Done():
+			if ctx.Err() == context.DeadlineExceeded {
+				t.Fatalf("expected p1 has the offset timelines [IDLE 1651161660600:106] -> [1651161660500:103] -> [-1:-1]..., [IDLE 1651161660600:106] -> [1651161660500:103] -> [-1:-1]..., "+
+					"[IDLE 1651161660600:106] -> [1651161660500:103] -> [-1:-1]... got %s %s %s: %s", allProcessors["p1"].GetOffsetTimelines()[0].Dump(), allProcessors["p1"].GetOffsetTimelines()[1].Dump(), allProcessors["p1"].GetOffsetTimelines()[2].Dump(), ctx.Err())
+			}
+		default:
+			time.Sleep(1 * time.Millisecond)
+			allProcessors = processorManager.GetAllProcessors()
+		}
+	}
+
+	// publish an idle watermark: simulate map
+	otValueByteOne, err = otValueToBytes(107, epoch+700, true, 0)
+	assert.NoError(t, err)
+	otValueByteTwo, err = otValueToBytes(107, epoch+700, true, 1)
+	assert.NoError(t, err)
+	otValueByteThree, err = otValueToBytes(107, epoch+700, true, 2)
+	assert.NoError(t, err)
+
+	err = otStore.PutKV(ctx, "p1", otValueByteOne)
+	assert.NoError(t, err)
+	err = otStore.PutKV(ctx, "p1", otValueByteTwo)
+	assert.NoError(t, err)
+	err = otStore.PutKV(ctx, "p1", otValueByteThree)
+	assert.NoError(t, err)
+
+	// p1 should get the head offset watermark from p2
+	for allProcessors["p1"].GetOffsetTimelines()[0].Dump() != "[IDLE 1651161660700:107] -> [IDLE 1651161660600:106] -> [1651161660500:103] -> [-1:-1] -> [-1:-1] -> [-1:-1] -> [-1:-1] -> [-1:-1] -> [-1:-1] -> [-1:-1]" &&
```

Comment:
Searching around our code base I notice that `GetOffsetTimelines` is used only by `GetHeadWatermark` and `GetWatermark`. It looks the `GetOffsetTimelines` is the implementation detail of `GetHeadWatermark` and `GetWatermark`, which should be hidden from unit tests. Unit tests should focus on test **what** instead of **how**. Do you think we can replace the `Dump() != "..."` check with `GetHeadWatermark() != "..."` or `GetWatermark() != "..."`?

Reply:
this is fine I guess.

Reply:
I have strong opinions on not exposing implementation details in unit tests. I approved the PR but we should address this later.